### PR TITLE
Tup build system, better errors, code restucture, new handlers, statistics & small fixes.

### PR DESCRIPTION
**Pull request by Jackojc**
- Code restructure
- CMake
- Better error reporting
- Sample data files

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -1,82 +1,305 @@
...
+                chr
+            };
         }
     );
 
 
 
 
+    // Identifiers.
+    ws::lexer::insert_case(match, "abcdefghijklmnopqrstuvwxyz",
```
Comment (Repertoi-e) [inline main.cpp]: Should identifiers be able to start with an underscore?

File: src/run/run.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+
+
+            // If we are on the final character,
+            // break so we don't get an exception from the next statement.
+            if (code.is_end())
+                break;
+
+
+            // Advance to the next character.
```
Comment (Repertoi-e) [inline src/run/run.cpp]:
```
// Increases i by one. 
++i;
```
Bad comment. Please avoid.

File: main.cpp (hunk 1 above)
Comment (Jackojc) [reply main.cpp]: Yeah, good point

File: src/run/run.cpp (hunk 2 above)
Comment (Jackojc) [reply src/run/run.cpp]: Lol

**Comment by Jackojc** [comment]
Implemented suggested improvements

**Comment by Jackojc** [comment]
Decided to switch to a new build system, please re-review.

**Comment by Jackojc** [comment]
Just gonna go ahead and merge this.

**Pull request merged**